gh-95411: IDLE - Enable using the module browser with .pyw files

erlend-aasland opened the pull request:
<!-- gh-issue-number: gh-95411 -->
* Issue: gh-95411
<!-- /gh-issue-number -->

erlend-aasland committed:
gh-89610: Don't hardcode supported extensions in IDLE browser

erlend-aasland committed:
List of extensions should be a set data type

erlend-aasland commented on Lib/idlelib/util.py:
```diff
@@ -13,8 +13,19 @@
+
 # .pyw is for Windows; .pyi is for stub files.
-py_extensions = ('.py', '.pyw', '.pyi')  # Order needed for open/save dialogs.
+py_extensions = {'.py', '.pyw', '.pyi'}  # Order needed for open/save dialogs.
```
I didn't have to change this, but a _set_ is a more correct data type for this application.

merwok replied:
Someone could argue that a tuple is a more useful type, as `py` is much more used than the others!
Also, the comment here explains another reason for order.

AlexWaygood replied:
Yes, in the initial version of my PR that added the `idlelib.util` submodule, I used a `frozenset` here. But @terryjreedy pointed out that the order of these extensions was important, and that therefore a `frozenset` would be unsuitable.

erlend-aasland replied:
No sweat. I'll revert the last commit.

erlend-aasland replied:
Resolved with 0c6e0687391744f1b51e5dff32343d97583fc032

erlend-aasland committed:
Revert "List of extensions should be a set data type"

This reverts commit a7757f7c4198ee19d51f17d6eef8ff97fb0cb9ae.

erlend-aasland committed:
Imports are sorted; insert new import at correct spot

erlend-aasland committed:
Also test extension is None

terryjreedy commented on Lib/idlelib/idle_test/test_util.py:
```diff
@@ -5,10 +5,26 @@
     def test_extensions(self):
-        for extension in {'.pyi', '.py', '.pyw'}:
+        for extension in self.supported_extensions:
             self.assertIn(extension, util.py_extensions)
```
When we added util.py, I merged with this test because we did not think of anything else.  Thinking more, making a duplicate set (in the math sense) makes it a 2nd source of truth, which seems wrong.  We should just import util.py_extensions and use it in other tests.

erlend-aasland replied:
IIUC, you're implying to remove `test_extensions` and keep the added tests?

erlend-aasland replied:
Resolved in 2041cca13457db0179e87aec3bc145cc60c097a2

terryjreedy reviewed (changes requested):
For me on main, .pyw can be browsed, .pyi cannot.  I am not surprised because .pyi files cannot be imported and it could be said that they are not modules.  In any case, pyclbr use importlib in `spec = importlib.util._find_spec_from_path(fullmodule, search_path)` and calls 
`spec.submodule_search_locations` and `spec.loader.get_source`.  We might have to continue to exclude .pyi files.

erlend-aasland committed:
Address review:

- use util.py_extensions in test
- remove now obsoleted UtilTest.test_extensions

erlend-aasland committed:
Address review: drop support for .pyi in browser

erlend-aasland committed:
Add NEWS

erlend-aasland committed:
Sync with main

erlend-aasland committed:
Add IDLE NEWS

terryjreedy commented on Lib/idlelib/util.py:
```diff
@@ -13,9 +13,31 @@
 py_extensions = ('.py', '.pyw', '.pyi')  # Order needed for open/save dialogs.
 
+# The browser depends on pyclbr and importlib which do not support .pyi files.
+browseable_extension_blocklist = ('.pyi',)
```
This file is for items used in multiple modules.  Please move this and the function to browser and adjust its imports.

Example of why: pyclbr and browser's use thereof might be changed so as to eliminate the limitation, at least when editing a .pyi file.  If so, it would be nice to only have to edit browser (separation of concerns).

erlend-aasland replied:
Yep, makes sense.

erlend-aasland replied:
BTW, the `is_supported_extension` is no longer in use. I'll remove it from the PR, unless you want it there for future use. Let me know what you prefer.

erlend-aasland replied:
Moved in 09b248d446e9a38847ecb36fadcf703dae4b840

erlend-aasland replied:
I reverted the now unneded changes to util.py in 04eef646d438c9c990a2be847789ef3827dc8330

erlend-aasland replied:
Marking this as resolved; please unresolve if you disagree.

terryjreedy commented on Misc/NEWS.d/next/IDLE/2022-07-29-11-08-52.gh-issue-95411.dazlqH.rst:
```diff
@@ -0,0 +1 @@
+Don't hardcode supported extensions in the IDLE browser.
```
```suggestion
Enable using IDLE's module browser with .pyw files.
```

terryjreedy commented on Lib/idlelib/NEWS.txt:
```diff
@@ -3,6 +3,8 @@ What's New in IDLE 3.11.0
 Released on 2022-10-03
 =========================
 
+gh-95411: Don't hardcode supported extensions in the IDLE browser.
```
```suggestion
gh-95411: Enable using IDLE's module browser with .pyw files.
```

terryjreedy reviewed (changes requested):
Pending a last look, I expect to merge after you move the browser stuff as requested above.

terryjreedy committed:
Update Misc/NEWS.d/next/IDLE/2022-07-29-11-08-52.gh-issue-95411.dazlqH.rst

terryjreedy committed:
Update Lib/idlelib/NEWS.txt

erlend-aasland committed:
Address review: move to browser

erlend-aasland committed:
Merge branch 'main' into idle/dont-hardcode-extensions

erlend-aasland committed:
Revert now unused changes in util.py

erlend-aasland committed:
Fix test
